### PR TITLE
Custom color for Keep

### PR DESCRIPTION
**Pull request by IgorOsiyuk**
Create palette component
image
set up animation for palette component 
set up functional for change keep's color 
update PUT-request for keeps

### CODE REVIEW

**Comment by melonholy** [comment]
Remove autocompletes 
image

File: google_keeps/src/components/PaletteComponent/PaletteForm.jsx [hunk 1]
```diff
@@ -1,21 +1,68 @@
...
-const PaletteForm = ({ changeColor, defaultColor }) => {
-  const [color, setColor] = useState(defaultColor);
+const PaletteForm = ({ changeColor, coordinate, closeModal }) => {
+  const [color, setColor] = useState();
+  const [isOutAnimate, setOutAnimate] = useState(zoomIn);
+  const paletteRef = useRef();
 
   useEffect(() => {
     changeColor(color);
   }, [color, changeColor]);
```
Comment (melonholy) [inline google_keeps/src/components/PaletteComponent/PaletteForm.jsx:39]: Remove use effecr

File: google_keeps/src/components/PaletteComponent/PaletteForm.jsx [hunk 2]
```diff
@@ -1,21 +1,68 @@
...
+  .react-colorful__pointer {
+    width: 20px;
+    height: 20px;
   }
 `;
 
-const PaletteForm = ({ changeColor, defaultColor }) => {
-  const [color, setColor] = useState(defaultColor);
+const PaletteForm = ({ changeColor, coordinate, closeModal }) => {
+  const [color, setColor] = useState();
```
Comment (melonholy) [inline google_keeps/src/components/PaletteComponent/PaletteForm.jsx:33]: remove local state

File: google_keeps/src/components/PaletteComponent/PaletteForm.jsx [hunk 3]
```diff
@@ -1,21 +1,68 @@
...
+  return createPortal(
+    <PalleteBox
+      tabIndex="0"
+      ref={paletteRef}
+      onBlur={onCloseModal}
+      coordinate={coordinate}
+      isOutAnimate={isOutAnimate}
+      onAnimationEnd={onAnimationEnd}
+    >
+      <PaletteBlock color={color} onChange={setColor} />
```
Comment (melonholy) [inline google_keeps/src/components/PaletteComponent/PaletteForm.jsx:62]: remove `color` prop and change `setColor` to `changeColor`

File: google_keeps/src/components/keeps/Keep.jsx [hunk 4]
```diff
@@ -114,10 +119,20 @@ const KeepControlButton = styled.button`
 `;
 
 const Keep = (props) => {
-  const { title, text, _id } = props.data;
+  const { title, text, _id, background } = props.data;
   const [keepData, setKeepData] = useState({ title, text });
+  const [bgColor, changeColor] = useState();
```
Comment (melonholy) [inline google_keeps/src/components/keeps/Keep.jsx:124]: set `background` prop as default

File: google_keeps/src/components/keeps/Keep.jsx [hunk 5]
```diff
@@ -114,10 +119,20 @@ const KeepControlButton = styled.button`
...
+
+  const formRef = useRef();
+  const refButton = useRef();
+
   const dispatch = useDispatch();
 
+  useEffect(() => {
+    formRef.current.style.backgroundColor = bgColor;
+  }, [bgColor]);
+
```
Comment (melonholy) [inline google_keeps/src/components/keeps/Keep.jsx:135]: remove use effect

File: google_keeps/src/components/keeps/Keep.jsx [hunk 6]
```diff
@@ -130,16 +145,30 @@ const Keep = (props) => {
...
+      background: bgColor,
+    };
+    dispatch(UpdateKeepRequest(_id, newData));
+  }, [_id, keepData, dispatch, bgColor]);
 
   const onDelete = useCallback(() => {
     dispatch(DeleteUserKeepRequest(_id));
   }, [_id, dispatch]);
 
+  const [isPalette, setPaletteState] = useState(false);
```
Comment (melonholy) [inline google_keeps/src/components/keeps/Keep.jsx:159]: group all set states in one place

File: google_keeps/src/components/keeps/Keep.jsx [hunk 7]
```diff
@@ -130,16 +145,30 @@ const Keep = (props) => {
...
+    setCoordinate({
+      x: refButton.current.getBoundingClientRect().x,
+      y: refButton.current.getBoundingClientRect().y,
+    });
+  }, [isPalette]);
+
   return (
     <KeepWrap>
-      <KeepForm>
+      <KeepForm bgColor={background} ref={formRef}>
```
Comment (melonholy) [inline google_keeps/src/components/keeps/Keep.jsx:171]: change `background` to `bgColor `

File: google_keeps/src/components/keeps/KeepsList.jsx [hunk 8]
```diff
@@ -6,6 +6,7 @@ const KeepListContainer = styled.div`
   flex-direction: row;
   flex-wrap: wrap;
   margin: 0 -16px;
+  position:relative;
```
Comment (melonholy) [inline google_keeps/src/components/keeps/KeepsList.jsx:9]: spaces)

**Review by melonholy — CHANGES_REQUESTED** [review]

File: google_keeps/src/components/PaletteComponent/PaletteForm.jsx (hunk 1 above)
Comment (IgorOsiyuk) [reply google_keeps/src/components/PaletteComponent/PaletteForm.jsx:39]: - [+]

File: google_keeps/src/components/PaletteComponent/PaletteForm.jsx (hunk 2 above)
Comment (IgorOsiyuk) [reply google_keeps/src/components/PaletteComponent/PaletteForm.jsx:33]: - [+]

File: google_keeps/src/components/PaletteComponent/PaletteForm.jsx (hunk 3 above)
Comment (IgorOsiyuk) [reply google_keeps/src/components/PaletteComponent/PaletteForm.jsx:62]: - [+]

File: google_keeps/src/components/keeps/Keep.jsx (hunk 4 above)
Comment (IgorOsiyuk) [reply google_keeps/src/components/keeps/Keep.jsx:133]: - [+]

File: google_keeps/src/components/keeps/Keep.jsx (hunk 5 above)
Comment (IgorOsiyuk) [reply google_keeps/src/components/keeps/Keep.jsx:144]: - [+]

File: google_keeps/src/components/keeps/Keep.jsx (hunk 6 above)
Comment (IgorOsiyuk) [reply google_keeps/src/components/keeps/Keep.jsx:168]: - [+]

File: google_keeps/src/components/keeps/Keep.jsx (hunk 7 above)
Comment (IgorOsiyuk) [reply google_keeps/src/components/keeps/Keep.jsx:180]: - [+]

File: google_keeps/src/components/keeps/KeepsList.jsx (hunk 8 above)
Comment (IgorOsiyuk) [reply google_keeps/src/components/keeps/KeepsList.jsx:8]: - [+]

File: google_keeps/src/Utils/hooks/debounce.jsx [hunk 9]
```diff
@@ -0,0 +1,13 @@
+import { useState, useEffect } from "react";
+export default function Debounce(value, delay) {
```
Comment (melonholy) [inline google_keeps/src/Utils/hooks/debounce.jsx:2]: name hooks like useDebouce

File: google_keeps/src/Utils/hooks/debounce.jsx [hunk 10]
```diff
@@ -0,0 +1,13 @@
...
+  const [debouncedValue, setDebouncedValue] = useState(value);
+  useEffect(() => {
```
Comment (melonholy) [inline google_keeps/src/Utils/hooks/debounce.jsx:4]: space between

File: google_keeps/src/Utils/hooks/debounce.jsx [hunk 11]
```diff
@@ -0,0 +1,13 @@
...
+    const handler = setTimeout(() => {
+      setDebouncedValue(value);
+    }, delay);
+    return () => {
```
Comment (melonholy) [inline google_keeps/src/Utils/hooks/debounce.jsx:8]: space

File: google_keeps/src/Utils/hooks/debounce.jsx [hunk 12]
```diff
@@ -0,0 +1,13 @@
...
+      clearTimeout(handler);
+    };
+  }, [delay, value]);
+  return debouncedValue;
```
Comment (melonholy) [inline google_keeps/src/Utils/hooks/debounce.jsx:12]: space

File: google_keeps/src/Utils/hooks/debounce.jsx (hunk 9 above)
Comment (melonholy) [reply google_keeps/src/Utils/hooks/debounce.jsx:2]: also name utils folder with small case

**Review by melonholy — CHANGES_REQUESTED** [review]

**Review by melonholy — APPROVED** [review]

**Pull request merged**